lib/gitgitgadget.ts

```diff
@@ -274,12 +274,10 @@ export class GitGitGadget {
         // Remove template from description (if template exists)
```

Comment:
I will change the patch to only touch `\r` vs `\n`.

> after we split by newlines, there is no Line Feed (\n) to be matched

I was puzzled by this myself but decided to restrain myself to apparent problems only.

> not you, not even me, knows what line ending convention the other users of GitHubGlue.getPRInfo() expect.

I did a search of `\r` everywhere and concluded that at best it's ignored. One other thing that I frankly didn't do is to search for references to `IPullRequestInfo.body`. I imagine that there's just one use for it.

> and merely replace the `\n?` by `\r?`

I considered that. However, there was other code that uses \n to append/concat lines again, producing a mixture of `/r` and `/r/n`. With other patch in mind that separates PR descriptions from other things, this is a smaller issue, yet still the footer parsing code will also (unexpectedly) replace `/r/n` with `/n`. I didn't like this idea and thought that it would be better to just get rid of `\r\n` early. What do you think?

Reply:
> > and merely replace the `\n?` by `\r?`
> 
> I considered that. However, there was other code that uses `\n` to append/concat lines, producing a mixture of `/n` and `/r/n`. With other patch in mind that separates PR descriptions from other things, this is a smaller issue, yet still the footer parsing code will also (unexpectedly) replace `/r/n` with `/n`. I didn't like this idea and thought that it would be better to just get rid of `\r\n` early. What do you think?

I think that if we really want to do a wholesale CR/LF -> LF, we have to be a lot more careful, and a PR that claims to fix the Cc: footers is probably the wrong venue for such an endeavor.

Reply:
OK.

Reply:
Shall I also concat lines with `\r\n` to preserve old logic?

Reply:
PS: If you like to "just do it yourself", feel free to throw away entire PR and reimplement. My desire to keep authorship only applied to super well known projects.

Reply:
`const match2 = line.match(/^([-A-Za-z]+:) (.*)\r?\n?$/);` does resolve the problem of only the last Cc being recognized.  At least it does on Windows, which is not where GitGitGadget runs.

Reply:
> If you like to "just do it yourself",

I prefer not to "just do it myself". That will require updates in this PR, though, and I have not seen any.

> feel free to throw away entire PR and reimplement.

I'd much rather not. But as I said, this would require changes in the branch to be force-pushed. The reviewer comments cost a lot of time, I only offer them when I think that something needs to be changed.

> My desire to keep authorship only applied to super well known projects.

I do not just take authorship. You are probably referring to our unfortunate encounter in https://github.com/gitgitgadget/git/pull/407, where I got so frustrated that I implemented the changes myself (which amounted to a complete rewrite, I did not even start from your patch, so I have no idea how you think that I "stole" the authorship from you. You claimed that you care a lot about authorship, so you should have been happy that I did not claim that you were the author of a patch that you did not, in fact, author).

I really hope that this PR does not go down the same route as https://github.com/gitgitgadget/git/pull/407 where you just "dug in" and refused to change anything according to my suggestions.

I really hope that this can be a regular open source PR where a contributor offers a useful change, reviewers help improve it, the contributor changes the commits accordingly, force-pushes, and the process repeats until everybody is happy and the PR is merged.

🤞

Reply:
I will apply the proposed changes to this PR, just give me a couple working days.

Reply:
I looked into the idea of changing regexp to `\r?` this today and it seems to be more complicated, in a way.

1. For simplicity, let's consider the other patch (that limits parsing to PR's body only) as already taken.
2. Now, the first regex in `parsePullRequestDescription()` is also doing something wrong: `/^([^]+)\n\n([^]+)$/` expects `\n\n` while it should expect `\r\n\r\n`. Let's assume that we also fix this.
3. Then, in order to preserve logic, we'll also need a change where unparsed footers are re-appended, there `\n\n${footer.join("\n")}` will also need to use `\r\n\r\n`.
4. Now, what if GitHub gives `\n` instead? That's already handled by `\r?`, but now `\r?` will need to be applied to first regex also,
5. Good, but now the re-appending is broken again. We'll need to capture `(\r?\n\r?\n)` into a variable and concat it later.

That's quite a bit of extra work.

Now, back to the idea of replacing `\r\n` early. I have studied the data flow and arrived at the following conclusions:
1. For coherence, `GitHubGlue.getOpenPRs` should also apply `\n` conversion. I overlooked this, but it seems to be harmless.
2. `IPullRequestInfo.body` is only used in `GitGitGadget.genAndSend()`.
3. There, I already handled `prBody` manipulation.
4. `prBody` only goes to `PatchSeries.getFromNotes()`.
5. `PatchSeries.getFromNotes()` calls `PatchSeries.parsePullRequestDescription()` which is fixed with my patches.
6. `PatchSeries.getFromNotes()` then gives result to `md2text()` which replaces `\r\n` with `\n` in `marked()`.

So the flow seems to be quite simple.

I would say that it's better to get rid of `\r` early and save quite a bit of code. With this information, what is your updated say on what to do?

Reply:
> 1. For simplicity, let's consider the other patch (that limits parsing to PR's body only) is already applied.
> 2. Now, the first regex in `parsePullRequestDescription()` is also doing something wrong: `/^([^]+)\n\n([^]+)$/` expects `\n\n` while it should expect `\r\n\r\n`. Let's assume that we also fix this.

I did not really see it as wrong.  My initial assumption was the `\n\n` was splitting the title from the body in `${pr.title}\n\n${prBody}${ccSubmitter}`.  I don't think users should have to know they need to leave a blank line before `Cc:`.

> 5. `PatchSeries.getFromNotes()` calls `PatchSeries.parsePullRequestDescription()` which is fixed with my patches.
> 6. `PatchSeries.getFromNotes()` then gives result to `md2text()` which replaces `\r\n` with `\n` in `marked()`.

> I would say that it's better to get rid of `\r` early and save quite a bit of code. With this information, what is your updated say on what to do?

So the change could be two bits?
1. GitHubGlue removes the `\r` 
2. `parsePullRequestDescription()` uses less greedy `/^([^]+)\n\n?([^]+)$/` to process all body lines.  As noted above I think all body lines should be processed (and they currently are).

Having said that, I have no problem with moving `pr.title` out of `parsePullRequestDescription()` and passing 'md2text(`${pr.title}\n\n${coverLetter}`' to the PatchSeries constructor.  Sorry about the formatting.

Reply:
Yes, I also thought that it would be reasonable to allow user to put `Cc:` anywhere and not require an empty line before it (that's kind of unexpected requirement). And yes, due to a bug, it always worked this way - except when it failed due to other bugs. So indeed maybe there's no reason to change that now. I only chose to fix (change?) code to parse "footer" because it looked like original code's intention.



> So the change could be two bits?
> 
>     1. GitHubGlue removes the `\r`
> 
>     2. `parsePullRequestDescription()` uses less greedy `/^([^]+)\n\n?([^]+)$/` to process all body lines.  As noted above I think all body lines should be processed (and they currently are).

If we choose to parse entire PR body, then with the other commit in mind, first regex `/^([^]+)\n\n([^]+)$/` can be dropped completely, because with the other commit `parsePullRequestDescription()` will only receive pr body and nothing else.

Reply:
It's a bit hard to judge what the code should do when it fails to work in so many ways (currently the only working case is when there's a single `Cc:` line in the very end without trailing newline).

If it was my choice, I would allow `Cc:` to be at any position and relax the requirement to have an empty line before the footer. That would be more reliable for users who didn't know (or forgot) about the "empty line" rule. The chance of false positive parsing is pretty low in my understanding.

Reply:
> If it was my choice, I would allow `Cc:` to be at any position and relax the requirement to have an empty line before the footer. That would be more reliable for users who didn't know (or forgot) about the "empty line" rule. The chance of false positive parsing is pretty low in my understanding.

No, it would not be reliable at all, and the chance of a false positive is very real. Take for example git/git@831a488b76e0fcad68c6d1c7aaf9f086199e1738: if the line

```
"Cc: Stable kernel <stable@vger.kernel.org> #v3.4 v3.5 v3.6"
```

would not have double-quotes around it, and was part of the PR description, it would be a false positive and the contributor would be rightfully annoyed. Or if the commit message used this formatting:

```
In some cases it is useful to add additional
information after the email address on the
Cc: footer in a commit log, for instance:

  "Cc: Stable kernel <stable@vger.kernel.org> #v3.4 v3.5 v3.6"
```

If you truly want to allow Cc:in anywhere in the PR description, you will have to make it a ton more robust, e.g. by requiring some sort of email address validation (which [is _hard_](https://changelog.com/posts/theres-only-one-way-to-validate-an-email-address) although we could go for ["good enough"](http://form.guide/best-practices/validate-email-address-using-javascript.html), and [thank goodness we don't have to do this in Perl](https://emailregex.com/)).

So if you want this, make sure to not only look for `Cc: ` at the beginning of the line (and [_make sure to do this case-insensitively_](https://github.com/git/git/pull/676#issuecomment-564007634)), but do verify that it is followed by a valid email address (or by a valid name plus email address in `<...>`, personally I would allow names that are all alpha-numerical characters plus `-` plus `.` plus double-quotes plus single-quotes plus single space, with at least one alphanumerical character in it), possibly followed by an arbitrary number of comma-separated recipients.

Reply:
> No, it would not be reliable at all, and the chance of a false positive is very real.
 
Thanks, that sounds like a valid concern. 

Alternatively, we could use some more unambiguous footers like `GitGitGadget-Cc:` but that would break existing habits, so probably not worth it either.

I will then keep with the idea to adjust the code so that only footer is parsed (which didn't work as expected before my patches).

Reply:
> No, it would not be reliable at all, and the chance of a false positive is very real.

This chance exists now due to `\n\n` splitting on title/body and not footer within body (due to `\r\n\r\n`.  I do not object to the concern - I did not realize there was a higher risk of it being in commit messages.